react/display-name false positive on ant-design ColumnsType

<!--
    ESLint adheres to the Open JS Foundation Code of Conduct:
    https://eslint.org/conduct

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://eslint.org/chat

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->


**Tell us about your environment**

<!--
    If you are using ESLint v6.5.0 or later, you can run ESLint with the `--env-info` flag and paste the output here.
-->

* **ESLint Version:** 7.5.0
* **ESLint Plugin React Version:** 7.20.3
* **Node Version:** 10.15.3
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    // Prettier plugin and recommended rules
    'prettier/@typescript-eslint',
    'plugin:prettier/recommended',
  ],
  rules: {
    // Include .prettierrc.js rules
    'prettier/prettier': ['error', {}, { usePrettierrc: true }],
    'react/prop-types': 'error',
    'react/react-in-jsx-scope': 'error',
    '@typescript-eslint/explicit-function-return-type': 'error',
    '@typescript-eslint/ban-ts-ignore': 'off',
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        labelComponents: [],
        labelAttributes: [],
        controlComponents: [],
        assert: 'either',
        depth: 25,
      },
    ],
    'jsx-a11y/anchor-is-valid': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'error',
    '@typescript-eslint/no-unused-vars': 'error',
  },
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
...
import { ColumnsType } from "antd/lib/table";
import { Table } from "antd";

const profilesColumns: ColumnsType<any> = [
  {
    title: "ID",
    dataIndex: "id",
    key: "id"
  },
  {
    title: "URL",
    dataIndex: "url",
    key: "url",
    render: (url: string): React.ReactElement => (
      <a href={url} target="_blank" rel="noopener noreferrer">
        <p>lol</p>
      </a>
    )
  }
];

ReactDOM.render(
  <Table columns={profilesColumns} />,
  document.getElementById("root")
);
```

<!-- Paste the command you used to run ESLint: -->
```bash
eslint --fix . --ext .ts,.tsx,.js,.jsx
```

**What did you expect to happen?**

For the Ant Design, `render` in `ColumnsType` is not a `React component` but function, hence the rule should not be applied in this scenario. No error should show up.

**What actually happened? Please include the actual, raw output from ESLint.**

```bash
yarn run v1.17.3
$ eslint --fix . --ext .ts,.tsx,.js,.jsx

/home/user/my-project/components/ComponentTable.tsx
  32:13  error  Component definition is missing display name  react/display-name

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -8,7 +8,11 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 ### Added
 * add [`no-namespace`] rule ([#2640] @yacinehmito @ljharb)
 
+### Fixed
+* [`display-name`]: Get rid of false position on component detection ([#2759] @iiison)
+
 [#2640]: https://github.com/yannickcr/eslint-plugin-react/pull/2640
+[#2759]: https://github.com/yannickcr/eslint-plugin-react/pull/2759
 
 ## [7.25.3] - 2021.09.19
 

--- a/lib/rules/display-name.js
+++ b/lib/rules/display-name.js
@@ -101,7 +101,7 @@ module.exports = {
       const namedFunctionExpression = (
         astUtil.isFunctionLikeExpression(node)
         && node.parent
-        && (node.parent.type === 'VariableDeclarator' || node.parent.method === true)
+        && (node.parent.type === 'VariableDeclarator' || node.parent.type === 'Property' || node.parent.method === true)
         && (!node.parent.parent || !utils.isES5Component(node.parent.parent))
       );
 
@@ -161,6 +161,7 @@ module.exports = {
         if (ignoreTranspilerName || !hasTranspilerName(node)) {
           return;
         }
+
         if (components.get(node)) {
           markDisplayNameAsDeclared(node);
         }

--- a/tests/lib/rules/display-name.js
+++ b/tests/lib/rules/display-name.js
@@ -504,6 +504,26 @@ ruleTester.run('display-name', rule, {
       class Link extends Component<LinkProps> {}
     `,
     parser: parsers.BABEL_ESLINT
+  }, {
+    code: `
+      const x = {
+        title: "URL",
+        dataIndex: "url",
+        key: "url",
+        render: url => (
+          <a href={url} target="_blank" rel="noopener noreferrer">
+            <p>lol</p>
+          </a>
+        )
+      }
+    `,
+    parser: parsers.BABEL_ESLINT
+  }, {
+    code: `
+      const renderer = a => function Component(listItem) {
+        return <div>{a} {listItem}</div>;
+      };
+    `
   }],
 
   invalid: [{
@@ -921,5 +941,12 @@ ruleTester.run('display-name', rule, {
       endLine: 11,
       endColumn: 11
     }]
+  }, {
+    code: `
+      const renderer = a => listItem => (
+        <div>{a} {listItem}</div>
+      );
+    `,
+    errors: [{message: 'Component definition is missing display name'}]
   }]
 });

